src/wallet/scriptpubkeyman.cpp

```diff
@@ -1434,3 +1496,609 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
+{
+    LOCK(cs_desc_man);
+    std::string error;
+    bool result = GetNewDestination(type, address, error);
+    index = descriptor.next_index - 1;
+    return result;
+}
+
+void DescriptorScriptPubKeyMan::ReturnDestination(int64_t index, bool internal, const CTxDestination& addr)
+{
+    LOCK(cs_desc_man);
+    // Only return when the index was the most recent
+    if (descriptor.next_index - 1 == index) {
+        descriptor.next_index--;
+    }
+    WalletBatch(m_storage.GetDatabase()).WriteDescriptor(GetID(), descriptor);
+    NotifyCanGetAddressesChanged();
+}
+
+std::map<CKeyID, CKey> DescriptorScriptPubKeyMan::GetKeys() const
+{
+    AssertLockHeld(cs_desc_man);
+    if (m_storage.HasEncryptionKeys() && !m_storage.IsLocked()) {
+        KeyMap keys;
+        for (auto key_pair : m_map_crypted_keys) {
+            const CPubKey& pubkey = key_pair.second.first;
+            const std::vector<unsigned char>& crypted_secret = key_pair.second.second;
+            CKey key;
+            DecryptKey(m_storage.GetEncryptionKey(), crypted_secret, pubkey, key);
+            keys[pubkey.GetID()] = key;
+        }
+        return keys;
+    }
+    return m_map_keys;
+}
+
+bool DescriptorScriptPubKeyMan::TopUp(unsigned int size)
+{
+    LOCK(cs_desc_man);
+    unsigned int target_size;
+    if (size > 0)
+        target_size = size;
+    else
+        target_size = std::max(gArgs.GetArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 0);
+    target_size = std::max((int32_t)target_size, 1);
+
+    // Calculate the new range_end
+    int32_t new_range_end = std::max(descriptor.next_index + (int32_t)target_size, descriptor.range_end);
+
+    // If the descriptor is not ranged, we actually just want to fill the first cache item
+    if (!descriptor.descriptor->IsRange()) {
+        new_range_end = 1;
+        descriptor.range_end = 1;
+        descriptor.range_start = 0;
+    }
+
+    FlatSigningProvider provider;
+    provider.keys = GetKeys();
+
+    WalletBatch batch(m_storage.GetDatabase());
+    uint256 id = GetID();
+    for (int32_t i = m_max_desc_index + 1; i < new_range_end; ++i) {
+        FlatSigningProvider out_keys;
+        std::vector<CScript> scripts_temp;
+        DescriptorCache cache;
+        if (!descriptor.descriptor->Expand(i, provider, scripts_temp, out_keys, &cache)) {
+            // Maybe we have a cached xpub and we can expand from cache
+            if (!descriptor.descriptor->ExpandFromCache(i, descriptor.cache, scripts_temp, out_keys)) return false;
+        }
+        // Add all of the scriptPubKeys to the scriptPubKey set
+        for (const CScript& script : scripts_temp) {
+            m_map_script_pub_keys[script] = i;
+        }
+        // Write the cache
+        std::map<KeyOriginInfo, CExtPubKey> newly_cached = descriptor.cache.GetNotCached(cache.GetCachedExtPubKeys());
```

Comment:
Now that you're no longer caching individual keys, can there still be a situation where you have a non-empty cache that's incomplete? If not, then you might be able to drop `GetNotCached`.

Reply:
Nvm, we still cache individual keys for `/*h` descriptors.

Reply:
Every `Expand` will also return the parent xpub. We don't want to be constantly rewriting that.

Reply:
Even if it was "unnecessary" I think it makes the code less brittle to future changes anyways.